### PR TITLE
Improve debug output and info logging on replace rewriter #181

### PR DESCRIPTION
**Pull request by TomGilke2468**
This solves issue 181.
image

### CODE REVIEW

File: querqy-core/src/main/java/querqy/rewrite/ContextAwareQueryRewriter.java [hunk 1]
```diff
@@ -1,39 +1,118 @@
...
  * method is called instead of {@link #rewrite(ExpandedQuery)} in the query rewriting chain. The Map argument can be
- * used to pass data between rewriters and to the consumer of the rewrite chain.
- * 
+ * used to pass data between rewriters and to the consumer of the rewrite chain. It also gives you the easy possibility to
+ * track the applied rules in a rewriter.
+ *
  * @author René Kriegler, @renekrie
  *
  */
-public interface ContextAwareQueryRewriter extends QueryRewriter {
```
Comment (renekrie) [inline querqy-core/src/main/java/querqy/rewrite/ContextAwareQueryRewriter.java:16]: Unfortunately this is too much of a change: turning the interface into an abstract class means that all custom rewriter implementations out there need to be changed.

How about this: leave ContextAwareQueryRewriter as an interface and create a new abstract class (AbstractLoggingRewriter??) that implements the interface and that looks more or less like your abstract ContextAwareQueryRewriter. You'd let the rewriters that come with Querqy implement it, and only those custom rewriters that use info logging need to be changed (which is probably very few)

File: querqy-core/src/main/java/querqy/rewrite/ContextAwareQueryRewriter.java [hunk 2]
```diff
@@ -1,39 +1,118 @@
...
      */
-    String CONTEXT_KEY_DEBUG_DATA = "querqy.debug.rewrite.data";
+    public static final String CONTEXT_KEY_DEBUG_DATA = "querqy.debug.rewrite.data";
+
+    /**
+     * Marker in the infologging for all applied rules.
+     */
+    private static final String APPLIED_RULES = "APPLIED_RULES";
+
+    protected SearchEngineRequestAdapter searchEngineRequestAdapter;
```
Comment (renekrie) [inline querqy-core/src/main/java/querqy/rewrite/ContextAwareQueryRewriter.java:45]: We should avoid making the SearchEngineRequest adapter a member variable. We have no guarantee that rewriter objects are created per request.

File: querqy-core/src/main/java/querqy/rewrite/ContextAwareQueryRewriter.java (hunk 1 above)
Comment (renekrie) [reply querqy-core/src/main/java/querqy/rewrite/ContextAwareQueryRewriter.java:16]: But see my note below re making the SearchEngineRequestAdapter a member variable.

**Review by renekrie — CHANGES_REQUESTED** [review]
Please see my comments re interface vs. abstract class and re making SearchEngineRequestAdapter a member variable. I haven't checked the changedsin the replace rewriter - will do, once the other issues have been ironed out.

File: querqy-core/src/main/java/querqy/rewrite/ContextAwareQueryRewriter.java (hunk 1 above)
Comment (TomGilke2468) [reply querqy-core/src/main/java/querqy/rewrite/ContextAwareQueryRewriter.java:16]: We (Tobi and I)  restored the interface and extracted the code to an abstract class like you suggested.

The SearchEngineRequestAdapter already was a member variable of the CommonRulesRewriter before so I moved it from the abstract class to both the CommonRules and the ReplaceRewriter.

What do you think about these changes?

**Review by renekrie — CHANGES_REQUESTED** [review]
Thank you, @TomGilke2468 ! I didn't see this in the first pass, but, unfortunately, the InfoLoggingContext also is a member variable and it also holds a reference to the SearchEngineRequestAdapter - so, same problem here. I think all you can do is, extract the utility methods to the AbstractLoggingRewriter (as you already did) and pass the SearchEngineRequestAdapter when you call these methods.

The thing is: it is the RewriterFactory that decided whether to create a new Rewriter instance per request or not. For the rewriters that come with Querqy, we also provide the factories and thus we have full control (that's why it was okay to have the SearchEngineRequestAdapter as a member var of CommonRulesRewriter). We don't have that control over custom rewriters.

**Review by renekrie — CHANGES_REQUESTED** [review]
Many thanks! Hopefully the last changes that I'm asking for:
* `appliedRules` is yet another member variable that could get us into trouble. I think you can just pass it to `rewriteContextAware()` (we could then even name this method back to `rewrite`)
* There is no test for the new InfoLogging and debug info in the ReplaceRewriter. Could you please add this? For inspiration: there is querqy.solr.InfoLoggingTest, which just uses the common rules rewriter. I suggest that you create a new test in the querqy.solr.rewriter.replace package.

**Comment by TomGilke2468** [comment]
Hi rene,
thanks for the suggestions on how to improve the code. :)

The newest commit addresses this remark:
> * `appliedRules` is yet another member variable that could get us into trouble. I think you can just pass it to `rewriteContextAware()` (we could then even name this method back to `rewrite`)

I kept the appliedRules as member variable of the CommonRulesRewriter to avoid having to make changes to the interface.

The tests are still a todo, hopefully I can get to that today.

**Review by renekrie — APPROVED** [review]

**Comment by renekrie** [comment]
Thank you @TomGilke2468. Fixes #181

**Pull request merged**